File: December2021/MiddleoftheLinkedList.java

class Solution {
	// TC : O(n)
	// SC: O(1)
	public ListNode middleNode(ListNode head) {
		if(head == null || head.next==null){
			return head;
		}

		ListNode fastPointer=head, slowPointer= head;
		while(fastPointer!=null && fastPointer.next!=null){
			fastPointer =fastPointer.next;
			if(fastPointer!=null){
				fastPointer =fastPointer.next;
			}
			slowPointer = slowPointer.next;

		}
		return slowPointer;
	}
}

File: December2022/C++/MiddleOfTheLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// Author: @mvssgupta

class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* slowPtr = head , *fastPtr = head;
        while(fastPtr && fastPtr->next)   
        {
            slowPtr = slowPtr->next;
            fastPtr = fastPtr->next->next;
        }
        return slowPtr;
    }
};
